### PR TITLE
add support for shared kind clusters

### PR DESCRIPTION
**Pull request by sdowell**
this change adds support to run kind clusters with a shared test environment. This can be run either sequentially or in parallel. The clusters are created at the beginning of the suite, reused between tests, and deleted at the end of the suite.

This is intended to:
- Speed up the presubmit jobs by reducing time spent creating kind clusters
- Make it easier to run tests with a shared test environment (e.g. reproducing test cleanup issues locally)
- Shift-left the shared environment testing to the presubmits so issues with test cleanup can be found before reaching periodics

It occurred to me while implementing this that we could use the same approach for the GKE testing, where we treat a set of GKE clusters as a shared pool in the tests. This approach could potentially replace the test group configuration, which would simplify the prow configuration. For now this does not make any changes to how GKE shared tests are run.

### CODE REVIEW

**Comment by sdowell** [comment]
Pulled out the kubeconfig changes as a separate PR: https://github.com/GoogleContainerTools/kpt-config-sync/pull/238

File: e2e/nomostest/testing/ntb.go
Comment (karlkfi) [inline e2e/nomostest/testing/ntb.go:106]: Does this mean we used to just skip all Cleanups on kind clusters?
That might explain why it was sometimes faster.

Comment (sdowell) [reply e2e/nomostest/testing/ntb.go:106]: Not exactly, this `FakeNTB` is used as an abstraction for shared test environments whose lifecycle is the entire test suite rather than per-test. Before this change we only really used this for GKE and any cleanup functions registered while creating the shared test environment would be ignored. There's a separate [Cleanup](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/e2e/nomostest/clean.go#L69) function that currently gets called for shared test environments on GKE.

The way we currently clean up the Kind clusters is by registering a cleanup function, so supporting cleanup functions with FakeNTB seemed like a straightforward solution to me (the only difference being FakeNTB cleanup gets invoked by us at the end of the test suite).

File: Makefile.e2e
Comment (sdowell) [reply Makefile.e2e:192]: This is the target used by the presubmit jobs, so my intent here was to make this the way we run presubmits. I haven't ran extensive profiling to measure the diff in runtime. The other (possibly more valuable) benefit is discovering issues with test cleanup before hitting the periodics.

I think [this make target](https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/Makefile.e2e#L209) probably serves the purpose of local development, although when I'm running these locally I find it a better experience to run these directly using `go test` rather than the make targets. These targets in particular I find pretty unfriendly to local usage since they create a container to run the tests in every time. 

Note the default when running the tests is to not use a shared test env (the `--share-test-env` flag must be provided). I don't feel too strongly about creating various permutations of make targets for the different flags, although at a certain point I feel it's simpler to just have developers run the tests directly with `go test`. Although we should probably do a better job at documenting all of the flag options in our dev guide

Comment (karlkfi) [reply Makefile.e2e:192]: Can you run some tests to make sure this is actually faster than the old way? I don't want to slow down pre-submits if we don't have to. We can always change this default behavior in a subsequent PR too.

File: e2e/nomostest/nt.go
Comment (nan-yu) [inline e2e/nomostest/nt.go:237]: If all `nt`s are in use, should we wait and retry acquiring an idle shared NT, instead of failing the test?

Comment (sdowell) [reply e2e/nomostest/nt.go:237]: This should never happen unless something else failed or there is a bug, because we create a number of shared test environments equal to the number of test threads. I figured it would be best to fail fast to make it obvious that something went wrong

Comment (nan-yu) [reply e2e/nomostest/nt.go:237]: ACK

**Review by nan-yu — APPROVED** [review]
/lgtm
/hold

LGTM, feel free to unhold once Karl's comments are addressed.

File: Makefile.e2e
Comment (karlkfi) [reply Makefile.e2e:192]: I'd love to use this for validating my finalizer PR, but I don't want to make it default without proof that it's not slower. If validating that will take a while, maybe just remove this line to avoid making this the default behavior and move that to another PR.

Comment (sdowell) [reply Makefile.e2e:192]: Sure that works for me, I can withhold updating the make target until we get a chance to profile it. You can see a few examples by looking at the current presubmit timings, but that's only a few data points.

From the current executions it seems to be fairly close to the [previous timings](https://oss.gprow.dev/?job=kpt-config-sync-presubmit-e2e-multi-repo-test-group3) (maybe a little faster), but this could probably be improved by optimizing some of the wait conditions during cleanup

Comment (sdowell) [reply Makefile.e2e]: Removed the change to the make target

/unhold

**Review by nan-yu — APPROVED** [review]
/lgtm

**Pull request merged**
